### PR TITLE
Add Mutable Reverb

### LINKED ISSUE #952: Port Clouds reverb model
Tracking issue for the Beads reverb model.

### PR DESCRIPTION
**Pull request by stellar-aria**
Closes #952

### CODE REVIEW

**Comment by soymonitus** [comment]
The new reverb model is astonishing. Thanks a lot.
I find though that changing Mode also resets the Roomsize, dampening and width. Is it really needed to reset them? If i have setup a "large" reverb with high values for those params, I would not want them to be changed if i change just the reverb model. Or is there any good reason to do it?

File: src/deluge/gui/ui/menus.cpp [hunk 1]
```diff
@@ -456,10 +427,12 @@ reverb::RoomSize reverbRoomSizeMenu{STRING_FOR_ROOM_SIZE};
 reverb::Dampening reverbDampeningMenu{STRING_FOR_DAMPENING};
 reverb::Width reverbWidthMenu{STRING_FOR_WIDTH, STRING_FOR_REVERB_WIDTH};
 reverb::Pan reverbPanMenu{STRING_FOR_PAN, STRING_FOR_REVERB_PAN};
+reverb::Model reverbModelMenu{STRING_FOR_MODEL};
 
 Submenu reverbMenu{
     STRING_FOR_REVERB,
     {
+        &reverbModelMenu,
```
Comment (seangoodvibes) [inline src/deluge/gui/ui/menus.cpp:435]: Can you add &reverbModelMenu to the `Submenu globalReverbMenu{` below as well :)? Thx!

**Review by seangoodvibes — CHANGES_REQUESTED** [review]

File: src/deluge/gui/ui/menus.cpp (hunk 1 above)
Comment (stellar-aria) [reply src/deluge/gui/ui/menus.cpp:435]: Done.

**Comment by soymonitus** [comment]
Small 'nit', in the global menu, the order is AMOUNT, MODEL, ROOM SIZE, while in the clip menu the order is MODEL, AMOUNT, ROOM SIZE. Could you make them consistent with each other?

File: src/deluge/gui/ui/menus.cpp [hunk 2]
```diff
@@ -536,6 +536,7 @@ Submenu globalReverbMenu{
     STRING_FOR_REVERB,
     {
         &globalReverbSendAmountMenu,
+		&reverbModelMenu,
```
Comment (seangoodvibes) [inline src/deluge/gui/ui/menus.cpp:539]: Small nit, as @soymonitus mentioned, can you make the reverbModelMenu first in the menu list.

So:

```
&reverbModelMenu,
&globalReverbSendAmountMenu,
```

Thx

**Review by seangoodvibes — CHANGES_REQUESTED** [review]

File: src/deluge/gui/ui/menus.cpp (hunk 2 above)
Comment (stellar-aria) [reply src/deluge/gui/ui/menus.cpp:539]: Original's model position has been changed to keep amount first

**Review by seangoodvibes — APPROVED** [review]
Thx for making the changes requested :)!

File: src/deluge/dsp/reverb/mutable/reverb.hpp [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+		for (size_t frame = 0; frame < in.size(); frame++) {
+			StereoSample& s = output[frame];
+			float wet = 0;
+			float apout = 0.0f;
+			engine_.Advance();
+
+			// Smear AP1 inside the loop.
+			// c.Interpolate(ap1, 10.0f, LFO_1, 80.0f, 1.0f);
+			// c.Write(ap1, 100, 0.0f);
```
Comment (sapphire-arches) [inline src/deluge/dsp/reverb/mutable/reverb.hpp:65]: Is this intentionally commented out?

File: src/deluge/dsp/reverb/mutable/reverb.hpp [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+	inline void set_time(float reverb_time) { reverb_time_ = reverb_time; }
+
+	inline void set_diffusion(float diffusion) { diffusion_ = diffusion; }
+
+	inline void set_lp(float lp) { lp_ = lp; }
+
+	inline void Clear() { engine_.Clear(); }
+
+	// Reverb Base Overrides
+	void setRoomSize(float value) override { reverb_time_ = 0.35f + 0.63f * value; }
```
Comment (sapphire-arches) [inline src/deluge/dsp/reverb/mutable/reverb.hpp:122]: style nit: we probably shouldn't use 3 different case styles for methods in the same class.

File: src/deluge/dsp/reverb/mutable/reverb.hpp (hunk 3 above)
Comment (stellar-aria) [reply src/deluge/dsp/reverb/mutable/reverb.hpp:65]: yes, see [Rings code](https://github.com/pichenettes/eurorack/blob/08460a69a7e1f7a81c5a2abcc7189c9a6b7208d4/rings/dsp/fx/reverb.h#L93)

File: src/deluge/dsp/reverb/mutable/reverb.hpp (hunk 4 above)
Comment (stellar-aria) [reply src/deluge/dsp/reverb/mutable/reverb.hpp]: Removed the unnecessary setters.

File: src/deluge/dsp/reverb/base.hpp [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+#include "dsp/stereo_sample.h"
+#include <cstdint>
+#include <span>
+
+namespace deluge::dsp::reverb {
+struct Base {
+	Base() = default;
+	virtual ~Base() = default;
+
+	virtual void Process(std::span<int32_t> input, std::span<StereoSample> output) = 0;
```
[the remark this answers is not in the document]
Comment (sapphire-arches) [reply src/deluge/dsp/reverb/base.hpp:11]: lowercase `process` would align better with the deluge code (both existing and our nominal style in clang-tidy)

[the remark this answers is not in the document]
Comment (stellar-aria) [reply src/deluge/dsp/reverb/base.hpp]: Done

**Pull request merged**